tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs

```diff
@@ -241,6 +242,172 @@ public async Task BlockPersistAndReverificationWillAbandonTxAsBalanceTransfered(
+            mp2.Attributes = new TransactionAttribute[] { new Conflicts() { Hash = mp1.Hash } };
+            _unit.TryAdd(mp2, engine.Snapshot);
+
+            var mp3 = CreateTransactionWithFeeAndBalanceVerify(2 * txFee);  // mp3 conflicts with mp1 and has larger network fee
+            mp3.Attributes = new TransactionAttribute[] { new Conflicts() { Hash = mp1.Hash } };
+            _unit.TryAdd(mp3, engine.Snapshot);
+
+            var mp4 = CreateTransactionWithFeeAndBalanceVerify(3 * txFee);  // mp4 conflicts with mp3 and has larger network fee
+            mp4.Attributes = new TransactionAttribute[] { new Conflicts() { Hash = mp3.Hash } };
+
+            var malicious = CreateTransactionWithFeeAndBalanceVerify(3 * txFee);  // malicious conflicts with mp3 and has larger network fee, but different sender
+            malicious.Attributes = new TransactionAttribute[] { new Conflicts() { Hash = mp3.Hash } };
+            malicious.Signers = new Signer[] { new Signer() { Account = new UInt160(Crypto.Hash160(new byte[] { 1, 2, 3 })), Scopes = WitnessScope.None } };
+
+            var mp5 = CreateTransactionWithFeeAndBalanceVerify(2 * txFee);  // mp5 conflicts with mp4 and has smaller network fee
+            mp5.Attributes = new TransactionAttribute[] { new Conflicts() { Hash = mp4.Hash } };
+
+            _unit.SortedTxCount.Should().Be(2);
+            _unit.UnverifiedSortedTxCount.Should().Be(0);
+
+            // Act & Assert: try to add conlflicting transactions to the pool.
+            _unit.TryAdd(mp1, engine.Snapshot).Should().Be(VerifyResult.HasConflicts); // mp1 conflicts with mp2 and mp3 but has lower network fee than mp3 => mp1 fails to be added
+            _unit.SortedTxCount.Should().Be(2);
+            _unit.GetVerifiedTransactions().Should().Contain(new List<Transaction>() { mp2, mp3 });
+
+            _unit.TryAdd(malicious, engine.Snapshot).Should().Be(VerifyResult.HasConflicts); // malicious conflicts with mp3, has larger network fee but malicious (different) sender => mp3 shoould be left in pool
+            _unit.SortedTxCount.Should().Be(2);
+            _unit.GetVerifiedTransactions().Should().Contain(new List<Transaction>() { mp2, mp3 });
+
+            _unit.TryAdd(mp4, engine.Snapshot).Should().Be(VerifyResult.Succeed); // mp4 conflicts with mp3 and has larger network fee => mp3 shoould be removed from pool
+            _unit.SortedTxCount.Should().Be(2);
+            _unit.GetVerifiedTransactions().Should().Contain(new List<Transaction>() { mp2, mp4 });
+
+            _unit.TryAdd(mp1, engine.Snapshot).Should().Be(VerifyResult.Succeed); // mp1 conflicts with mp2 and has same network fee, mp2 has Conflicts attribute => mp2 should be removed from pool
+            _unit.SortedTxCount.Should().Be(2);
+            _unit.GetVerifiedTransactions().Should().Contain(new List<Transaction>() { mp1, mp4 });
```

Comment:
@AnnaShaleva and @superboyiii,
I do not fully understand this case.

`mp2` has the `conflicts` attribute with `mp1`.
Initially, I had understood that `mp1` was a transaction that was send first and then the user decided to replace it with mp2. Why the opposite had occur?

Reply:
This testcase is testing another scenario: the testcase assumes that `mp2` is *already* in the pool (e.g. it was received earlier from some other node), and after that the node receives `mp1`. `mp1` then will replace `mp2`, because they have the same network fee and `mp1` doesn't have the `Conflicts` attribute with `mp2`'s hash.

> `mp1` was a transaction that was send first and then the user decided to replace it with `mp2`
We have another testcase for that scenario, see the following:
```
            _unit.TryAdd(mp2_1, engine.Snapshot).Should().Be(VerifyResult.HasConflicts); // mp2_1 conflicts with mp1 and has same network fee, mp2_1 has Conflicts attribute => mp2_1 shoouldn't be added to the pool
            _unit.SortedTxCount.Should().Be(2);
            _unit.GetVerifiedTransactions().Should().Contain(new List<Transaction>() { mp1, mp4 });
```
In this testcase `mp1` is already in the mempool and the user tries to replace it with `mp2_1` that has the same network fee and `Conflicts` attribute with the hash of `mp1`. The user will fail to replace it because `mp2_1` has the same network fee. If the user wants to replace transaction that exists in the mempool, then the user have to pay more.

Reply:
I understand. 

That makes sense, thanks for the careful explaining @AnnaShaleva .

Just to clarify.
In the case that mp2 arrives in a node and it proposes a block, maybe the `backups`(other CN), will have mp1 first and won't be able to obtain mp2, right?

The view will be changed.

Reply:
> In the case that mp2 arrives in a node and it proposes a block, maybe the backups(other CN), will have mp1 first and won't be able to obtain mp2, right?

No, I don't think so. In this case the backups will request the proposal and find out that `mp2` is missing from the dBFT context. After that I'd expect them to request missing `mp2` from their neighbours (and, likely, from the primary that has `mp2` or maybe from some other nodes that also still have `mp2` in the pool). After they get `mp2`, they will send it directly to the dBFT context bypassing the mempool (i.e. the `mp2` *will NOT be rejected* due to conflicts). Thus, I'd expect the block to be accepted. No change view will happen (if the timer won't fire earlier than all backups get the missing `mp2`).

Reply:
If they have `mp1` already in their mempool I am understading that `mp2` will fail to be verified

Reply:
I see what you mean. Maybe it is bypassed.

Reply:
From my side, I believe that almost all attack vector our team had thought were covered.

However, the recent issue I mentioned above opened a new space for discussion in my opinion.

Reply:
> If they have mp1 already in their mempool I am understading that mp2 will fail to be verified

Only in case if we're trying to add  `mp2` directly to the mempool, but...

> Maybe it is bypassed

As far as I understand the C# node code, the blockchain router receives the transaction, verifies state independant checks and then passes it to the subscribers. And the DBFT plugin receives it directly from the router (not after mempool). Is it true?

Reply:
> It has an isolated context.

That's definitely true, this context contains the proposed transactions for the current block and if the block is correct, then `mp2` is allowed in this context.